Add TMTensorToLinalgExtPass

cathyzhyi opened the pull request:
This PR adds the conversions from `tm_tensor` to `linalg_ext`ops. The goal is to use `tm_tensor` as an interface dialect to bridge between projects. The `tm_tensor` dialect will be upstreamed eventually. A high level description about `tm_tensor` dialect can be found in
torch-mlir-dialects/include/torch-mlir-dialects/Dialect/TMTensor/IR/TMTensorBase.td

The change are broken up in to 3 commits:
1.  Copy over torch-mlir-dialects from https://github.com/llvm/torch-mlir without any changes.
2.  Hook up the tmtensor dialect in third_party/torch-mlir-dialects with new BUILD.overlay file, register dialects and update the third_party/torch-mlir-dialects/README.md to include instructions on how to pull in torch-mlir-dialects from upstream.
3.  Add `TMTensorToLinalgExtPass` which is a 1 to 1 mapping
from `tm_tensor` ops to `linalg.ext` ops. The syntax of the ops from the
two dialects are identical. Add --iree-input-type=tmtensor input type.

GMNGeoffrey commented:
Is this ready for review or still in progress? Could you please add a description? This is a lot of code to have no description :-)

cathyzhyi commented:
> Is this ready for review or still in progress? Could you please add a description? This is a lot of code to have no description :-)

Sry, I just added the description. Yea, it's ready for review though I am still fixing errors reported from the bazel build.

MaheshRavishankar commented on llvm-external-projects/torch-mlir-dialects/include/torch-mlir-dialects/Dialect/TMTensor/IR/TMTensorInterfaces.h:
```diff
@@ -0,0 +1,42 @@
+class TMTensorOp;
+
+/// OpOperand vector that implicitly converts to a Value vector.
+struct OpOperandVector : public SmallVector<OpOperand *> {
```
I dont think we need this. Such a construct exists in core already?

cathyzhyi replied:
replaced with the one from core. Thanks!

cathyzhyi commented on iree/compiler/InputConversion/TMTensor/ConvertTMTensorToLinalgExt.cpp:
```diff
@@ -0,0 +1,90 @@
+namespace {
+
+template <typename SrcOpTy, typename TargetOpTy>
+struct TMTensorOpConversion : public OpConversionPattern<SrcOpTy> {
```
removed. Thanks!

cathyzhyi commented on iree/compiler/Pipelines/Pipelines.cpp:
```diff
@@ -45,6 +46,8 @@ void buildIREEVMTransformPassPipeline(
       MHLO::buildMHLOInputConversionPassPipeline(passManager);
       break;
   }
+  passManager.addNestedPass<FuncOp>(
```
Thanks for the note. There are no more passes AFAICT. Will add the new input type if things change.

cathyzhyi commented on llvm-external-projects/torch-mlir-dialects/include/torch-mlir-dialects/Dialect/TMTensor/IR/TMTensorInterfaces.td:
```diff
@@ -0,0 +1,493 @@
+        uninitialized, as well as properly cloning "read-write" operands.
+      }],
+      /*retTy=*/"bool",
+      /*methodName=*/"payloadUsesValueFromOperand",
```
hmm, this is currently used in the bufferization pass to query if the initial value is used and need to copy over to the new buffer allocated. 
Called in bufferization pass for torch-mlir reference backend: 
https://github.com/llvm/torch-mlir/blob/5b9bdfaf3fb56c5aed12d689e21c9884b1c5a789/external/llvm-external-projects/torch-mlir-dialects/lib/Dialect/TMTensor/Transforms/Bufferize.cpp#L63
The implementation for tmtensor ops:
scan op: https://github.com/llvm/torch-mlir/blob/5b9bdfaf3fb56c5aed12d689e21c9884b1c5a789/external/llvm-external-projects/torch-mlir-dialects/lib/Dialect/TMTensor/IR/TMTensorOps.cpp#L169
scatter op: https://github.com/llvm/torch-mlir/blob/5b9bdfaf3fb56c5aed12d689e21c9884b1c5a789/external/llvm-external-projects/torch-mlir-dialects/lib/Dialect/TMTensor/IR/TMTensorOps.cpp#L401

I guess when this is upstreamed the bufferization pass need to be changed to support the new bufferization and this might not be needed eventually.

MaheshRavishankar replied:
That might be a bug... The method in `Linalg` ops has a more defined meaning. That meaning does not translate over to these ops. The region definition is more like a "binary operator". So if something is using `payloadUsesValueFromOperand` it might be using it the wrong way.

cathyzhyi replied:
Got it. Will remove the API. Thanks!

cathyzhyi commented on llvm-external-projects/torch-mlir-dialects/include/torch-mlir-dialects/Dialect/TMTensor/IR/ScalarLoopOpInterface.td:
```diff
@@ -0,0 +1,79 @@
+          the `dest` parameter.
+        }],
+        /*retType=*/"SmallVector<Value>",
+        /*methodName=*/"getDestinationOperands",
```
removed. thanks!

cathyzhyi commented on llvm-external-projects/torch-mlir-dialects/lib/Dialect/TMTensor/Transforms/Bufferize.cpp:
```diff
@@ -0,0 +1,156 @@
+//===- Bufferize.cpp - Bufferization of tmtensor ops ------------------===//
```
I am actually moving `torch-mlir-dialects` to `third_party` as suggested by Stella. Then it's more obvious this is code from third party. Would it be OK then if it contains code not used by IREE?

MaheshRavishankar replied:
It adds some maintanence burden cause the code needs to be kept in sync with LLVM. I'd rather only take those files that are needed for the conversion to `LinalgExt`

stellaraccident replied:
If the build rules can be sliced so that we only actually depend on the dialect, then the extra files should not be a problem (ie. Allowing more of a straight CP). It would be better though if the upstream source were factored to be really thin, just providing the dialect.

If just the dialect, feel free to just include it in the main iree-opt and have a really simple smoke test along with the transform into IREE.

cathyzhyi replied:
> It would be better though if the upstream source were factored to be really thin, just providing the dialect.

Do you mean separate `lib/Dialect/TMTensor/IR (for dialects)` and `lib/Dialect/TMTensor/Transforms (for transforms)` into their own projects and include the dialects part for IREE?

MaheshRavishankar replied:
We probably dont need the entire `TMTensor` in IREE. We probably just need `TMTensor/IR`, and the pass here to convert from `TMTensor` to `LinalgExt`?

cathyzhyi replied:
got it. Removed everything else and only keep the TMTensor/IR.

cathyzhyi commented on llvm-external-projects/torch-mlir-dialects/tools/torch-mlir-dialects-opt/torch-mlir-dialects-opt.cpp:
```diff
@@ -0,0 +1,49 @@
+//===------------------------------------------------------------*- C++ -*-===//
```
Removed the tools directory as it's not needed by IREE.

cathyzhyi commented on iree/tools/ireec_lib.cc:
```diff
@@ -82,6 +83,7 @@ int mlir::iree_compiler::runIreecMain(int argc, char **argv) {
   mlir::iree_compiler::registerAllPasses();
   mlir::iree_compiler::registerIreeDialects(registry);
   mlir::iree_compiler::registerIreeCompilerModuleDialects(registry);
+  registry.insert<mlir::torch::TMTensor::TMTensorDialect>();
```
Done!

cathyzhyi replied:
Done!

ScottTodd commented on third_party/torch-mlir-dialects/README.md:
```diff
@@ -0,0 +1,11 @@
+# Torch-MLIR Dialects Project
+
+Sources for torch-mlir's public dialects (containing ops/types/attributes that
+are unique to Torch-MLIR at the moment)
+
+This project is intended to be used via LLVM's external projects setup:
+
+* `-DLLVM_EXTERNAL_PROJECTS=torch-mlir-dialects`
+* `-DLLVM_EXTERNAL_TORCH_MLIR_DIALECTS_SOURCE_DIR={this_directory}`
+
+It depends on the `mlir` project.
```
So these files are coming from https://github.com/llvm/torch-mlir? Please include that information in the README here, and also include instructions for updating these files (which directories to copy, which files to exclude).

Could we use a git submodule instead? Most of https://github.com/google/iree/tree/main/third_party is set up that way, though large optional deps that don't often require local changes like TensorFlow and Tint are fetched on demand by the build system:

https://github.com/google/iree/blob/47e45d3c433a4b20eb9ef28c2340e47ae7108ca0/integrations/tensorflow/WORKSPACE#L12-L16

https://github.com/google/iree/blob/47e45d3c433a4b20eb9ef28c2340e47ae7108ca0/build_tools/third_party/tint/CMakeLists.txt#L30-L31

How frequently will these files need to be updated? Will the files only be updated upstream then pulled here, or will local modifications be needed?

cathyzhyi replied:
Hey Scott, we didn't go with the submodule approach because this is a temporary solution and will all be removed once the tmtensor dialect is upstreamed to core which is expected to happen in Q2.

> How frequently will these files need to be updated? 

I don't expect the files would be updated very often if at all. Technically, this depends on what LinalgExt ops are needed to run the targeted torch models or if LinalgExt ops (scan and scatter op for now) interface changes in IREE. The workflow is a bit weird at this point: `a LinalgExt op is needed or its interface changed -> copy the op ods to torch-mlir-dialects as a tmtensor op -> copy torch-mlir-dialects to iree`. However, the scatter op is the only one we are aware of that's actually needed in the near future. 

> Will the files only be updated upstream then pulled here, or will local modifications be needed?

The files should be updated upstream and then pulled here 
- with some directories removed like `test`, `tools`, `lib/Dialect/TMTensor/Transforms`, `include/torch-mlir-dialects/Dialect/TMTensor/Transforms`. 
- with format changed according to IREE's standard. Torch-mlir uses different  `clang-format`.
- llvm version bump: torch-mlir will try best to keep up with llvm version used by IREE. However,  it should be OK to apply changes related to llvm version bumping to torch-mlir-dialects.

ScottTodd replied:
> we didn't go with the submodule approach because this is a temporary solution and will all be removed once the tmtensor dialect is upstreamed to core which is expected to happen in Q2

Files exist in the project's git history forever, while submodules are pretty cheap to add and remove, though they do take time to fetch from when they exist. @GMNGeoffrey any thoughts on this?

> * with format changed according to IREE's standard. Torch-mlir uses different clang-format.

we explicitly disable formatting in third_party/, and pristine copies of files are much easier to work with, so you can remove that step

https://github.com/google/iree/blob/47e45d3c433a4b20eb9ef28c2340e47ae7108ca0/third_party/.clang-format#L15-L17

ScottTodd replied:
Submodules don't _need_ to be narrow projects, but that can be nice when possible. For example, https://github.com/tensorflow/mlir-hlo is a standalone repository mirroring the [tensorflow/compiler/mlir/hlo](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/compiler/mlir/hlo) folder in tensorflow's full repo. On the other hand, we include all of [llvm/llvm-project](https://github.com/llvm/llvm-project/) as a submodule.

Another benefit to using submodules is that they can be updated in the same way (e.g. https://stackoverflow.com/a/5828396)

MaheshRavishankar replied:
I dont know if `torch-mlir` can be made a submodule of IREE. Torch-mlir uses IREE and this will add a cyclic dependence. The only way to break the dependence is to move the parts that are needed to connect to IREE into its own repository from where Torch and IREE use it. Kind of like MLIR-HLO.
Better voices might make a more reasoned decision here. When discussing with Stella, she suggesting just copying things instead of submodules. If you ask me, I will go with what Stella says, but you guys know better than I do.

ScottTodd replied:
> Torch-mlir uses IREE and this will add a cyclic dependence.

This no longer appears to be the case as of https://github.com/llvm/torch-mlir/commit/1a0b953ea7edfaf26fd3881d102c64fe0c1aab29?

Cycles are also only an issue if you recursively fetch submodules, which we discourage. That is a good point for projects downstream from IREE though (they may opt for a recursive submodule fetch, which we don't want to break).

> Though not sure how this would work with LLVM bumps.

This has to be solved regardless of how the code exists in the repo. If we fork the code into IREE (as this PR does) then we'd need to maintain that fork directly. If we pull a submodule then we need to update that submodule as we do our other LLVM-dependent submodules.

cathyzhyi replied:
How does the llvm bump for other submodule projects in IREE work? If IREE wants to bump llvm to a specific version but a submodule project is not at that version yet would IREE requires the upstream project to bump and then pull in the changes? or someone from IREE needs to fix all the related code for the submodule within IREE?

ScottTodd replied:
Yep, since LLVM and MLIR offer no compatibility guarantees. We have https://github.com/google/iree-mhlo-fork, https://github.com/google/iree-llvm-fork, and https://github.com/google/iree-tf-fork in case we need to carry forward local fixes. Each dependency increases this cost.

As it relates to this new dependency, we have a few options:

1) Fork the code into third_party. Start with a pristine copy then include local changes as part of bumping the LLVM version. These diffs will be difficult to contribute back to upstream and diverging will be costly if development continues in two locations.
2) Add a direct submodule and require that it move in lock-step with the rest of IREE's deps (mostly driven by the opaque Google LLVM integrate).
3) Add a forked submodule and hope for it to move in lock-step but allow patches in the fork as needed.

Isolating this dependency (and the others) can help if compatibility becomes difficult to maintain. For example, we could put deps in [experimental](https://github.com/google/iree/tree/main/experimental) or behind a CMake option so they are easy to disable temporarily.

GMNGeoffrey replied:
Generally all the projects we depend on that depend on LLVM are kept at the bleeding edge because they're being actively developed alongside MLIR. We have fork repositories of MHLO, TF, and LLVM, so that we can add local patches if we need to. Unclear if that would be necessary here. Generally we would wait for our dependencies to bump. If we're just copying a single file, then I think it makes sense to copy it into third_party. But if we're copying a significant chunk of the repo, as appears to be the case here, I would strongly prefer a submodule. One cost of a submodule is that when it's checked out, you're checking out the entire repo (not just the files you need), but an advantage is that the only thing stored in the git history is a commit hash. I would therefore say that if something is temporary that's even *more* reason to use a submodule rather than copying files.

MaheshRavishankar commented on iree/compiler/InputConversion/TMTensor/ConvertTMTensorToLinalgExt.cpp:
```diff
@@ -0,0 +1,87 @@
+    target.addIllegalDialect<mlir::torch::TMTensor::TMTensorDialect>();
+    target.addLegalDialect<IREE::LinalgExt::IREELinalgExtDialect>();
+
+    if (failed(applyPartialConversion(getOperation(), target,
```
This needs to be `applyPatternsAndFoldGreedily`

MaheshRavishankar reviewed (changes requested):
This is much better. Thanks. Still a comment on the conversion to LinalgExt. It doesnt need the dialect conversion infra.

MaheshRavishankar commented:
@cathyzhyi couple of things
1) Could you please wait for a bit to land this PR. I'd be really helpful to land this after the next IREE integrate that has some very significant changes from upstream to account for. If you could help by rebasing this PR after the integrate lands, that would be appreciated.
2) Also there are no lit-tests for `TMTensor`. If there are IR tests for `TMTensor` IR maybe those are OK to add. Will give us an indication of something broken and we can address that (will probably reach out to you for involved cases anyway), but at least they wont be silent failures/untested code within IREEs codebase.

cathyzhyi commented on iree/compiler/InputConversion/TMTensor/ConvertTMTensorToLinalgExt.cpp:
```diff
@@ -0,0 +1,87 @@
+
+#undef INSERT_TMTENSOR_CONVERSION_PATTERN
+
+    target.addIllegalDialect<mlir::torch::TMTensor::TMTensorDialect>();
```
sry, missed this. Done!

cathyzhyi commented:
@MaheshRavishankar Thanks a lot for reviewing!

yea, I can wait until next IREE integrate. Please let me know when it's done or which PR or issue I can watch for.

> Also there are no lit-tests for TMTensor. If there are IR tests for TMTensor IR maybe those are OK to add. Will give us an indication of something broken and we can address that (will probably reach out to you for involved cases anyway), but at least they wont be silent failures/untested code within IREEs codebase.

The TMTensor lit-tests are using torch-mlir-dialect-opt binary which is removed. Is it ok to add those tests to iree/compiler/InputConversion/TMTensor/test/ using iree-opt?

MaheshRavishankar commented:
> @MaheshRavishankar Thanks a lot for reviewing!
> 
> yea, I can wait until next IREE integrate. Please let me know when it's done or which PR or issue I can watch for.
> 

This is the integrate PR https://github.com/google/iree/pull/8612 . 

> > Also there are no lit-tests for TMTensor. If there are IR tests for TMTensor IR maybe those are OK to add. Will give us an indication of something broken and we can address that (will probably reach out to you for involved cases anyway), but at least they wont be silent failures/untested code within IREEs codebase.
> 
> The TMTensor lit-tests are using torch-mlir-dialect-opt binary which is removed. Is it ok to add those tests to iree/compiler/InputConversion/TMTensor/test/ using iree-opt?

You should be able to keep those in `iree-dialects/test/...` and use `iree-opt` ?

cathyzhyi commented:
> You should be able to keep those in iree-dialects/test/... and use iree-opt?

yea, I can do that. Just need an extra step to change the build files and the .mlir file manually when cp & paste from torch-mlir

GMNGeoffrey commented:
> > You should be able to keep those in iree-dialects/test/... and use iree-opt?
> 
> yea, I can do that. Just need an extra step to change the build files and the .mlir file manually when cp & paste from torch-mlir

This again points to using a submodule to me. We can run a dependent projects' tests from their repository. We shouldn't need to be copying them over with modifications to use our binaries, etc etc.

silvasean commented:
Just to give a bit more context here, the TMTensor dialect is basically an "interface dialect" mirroring linalg_ext. This dialect itself just exists transiently when IR is in textual asm syntax between projects, so it doesn't really have a super strict commit hash compatibility requirement (the syntax is fairly stable). We have a plan to upstream linalg_ext in some form so that this dialect isn't needed at all -- @MaheshRavishankar can you describe the plan?

cathyzhyi commented:
I've addressed all the comments and rebased on the latest. There are two issues left:

1. [bazel tset](https://source.cloud.google.com/results/invocations/7d788f89-3f09-49ca-a995-274629a23e37/targets/iree%2Fgcp_ubuntu%2Fbazel%2Flinux%2Fx86-swiftshader%2Fcore%2Fpresubmit/log) failed because `iree/compiler/InputConversion/TMTensor/BUILD:64:11: no such package 'third_party/torch-mlir-dialects': Package is considered deleted due to --deleted_packages and referenced by '//iree/compiler/InputConversion/TMTensor:init_tm_tensor_dialect'` Any suggestions on how to fix this?

3. I added a new third_party/torch-mlir-dialects/test/CMakeLists.txt for the IR test to use `opt-iree` but not sure if it's added correctly.

MaheshRavishankar reviewed (approved):
Thanks! These are good to go now. THanks for adding the tests.

MaheshRavishankar commented:
> Just to give a bit more context here, the TMTensor dialect is basically an "interface dialect" mirroring linalg_ext. This dialect itself just exists transiently when IR is in textual asm syntax between projects, so it doesn't really have a super strict commit hash compatibility requirement (the syntax is fairly stable). We have a plan to upstream linalg_ext in some form so that this dialect isn't needed at all -- @MaheshRavishankar can you describe the plan?

Missed this notification. Yes, these ops will move upstream. The plan is a bit involved, but I expect to have an RFC in a few weeks. If more detail needed here I can go into more details about the plan as of now.

MaheshRavishankar commented:
> > You should be able to keep those in iree-dialects/test/... and use iree-opt?
> 
> yea, I can do that. Just need an extra step to change the build files and the .mlir file manually when cp & paste from torch-mlir

You could probably just write a simple script to change the binary used during copy?

Thanks for adding the tests. If you have some roundtrip tests you want to copy, those would be a good idea as well.

MaheshRavishankar commented on third_party/torch-mlir-dialects/BUILD:
```diff
@@ -0,0 +1,151 @@
+load("@llvm-project//mlir:tblgen.bzl", "gentbl_cc_library", "td_library")
```
Actually I am confused why these are in `third_party/...`. Ill reach out offline.

MaheshRavishankar reviewed (changes requested):
Just noticed something in third_party/... . Not sure why?

MaheshRavishankar commented:
> I've addressed all the comments and rebased on the latest. There are two issues left:
> 
> 1. [bazel tset](https://source.cloud.google.com/results/invocations/7d788f89-3f09-49ca-a995-274629a23e37/targets/iree%2Fgcp_ubuntu%2Fbazel%2Flinux%2Fx86-swiftshader%2Fcore%2Fpresubmit/log) failed because `iree/compiler/InputConversion/TMTensor/BUILD:64:11: no such package 'third_party/torch-mlir-dialects': Package is considered deleted due to --deleted_packages and referenced by '//iree/compiler/InputConversion/TMTensor:init_tm_tensor_dialect'` Any suggestions on how to fix this?
> 2. I added a new third_party/torch-mlir-dialects/test/CMakeLists.txt for the IR test to use `opt-iree` but not sure if it's added correctly.

Must have missed the conversation related to this, but putting this in `llvm-external-project/iree-dialects` should solve these issues.

cathyzhyi commented:
> > I've addressed all the comments and rebased on the latest. There are two issues left:
> > 
> > 1. [bazel tset](https://source.cloud.google.com/results/invocations/7d788f89-3f09-49ca-a995-274629a23e37/targets/iree%2Fgcp_ubuntu%2Fbazel%2Flinux%2Fx86-swiftshader%2Fcore%2Fpresubmit/log) failed because `iree/compiler/InputConversion/TMTensor/BUILD:64:11: no such package 'third_party/torch-mlir-dialects': Package is considered deleted due to --deleted_packages and referenced by '//iree/compiler/InputConversion/TMTensor:init_tm_tensor_dialect'` Any suggestions on how to fix this?
> > 2. I added a new third_party/torch-mlir-dialects/test/CMakeLists.txt for the IR test to use `opt-iree` but not sure if it's added correctly.
> 
> Must have missed the conversation related to this, but putting this in `llvm-external-project/iree-dialects` should solve these issues.

yea, Stella suggested moving under third_party. @ScottTodd Do you have any suggestions about this?

MaheshRavishankar reviewed (approved):
Dropping my blocker for some of this being in `third_party/...`

GMNGeoffrey commented on build_tools/bazel_to_cmake/bazel_to_cmake_targets.py:
```diff
@@ -22,6 +22,9 @@
     "//llvm-external-projects/iree-dialects:IREELinalgExtPasses": [
         "IREELinalgExtPasses"
     ],
+    "//third_party/torch-mlir-dialects:TorchMLIRTMTensorDialect": [
```
This needs to be set up as a Bazel external repository. Is the BUILD file there copied from upstream? If not, we should put it in the build tools third party overlay directory. Here's an example:

https://github.com/google/iree/blob/47e45d3c433a4b20eb9ef28c2340e47ae7108ca0/build_tools/bazel/workspace.bzl#L52-L57

```suggestion
    "@torch-mlir-dialects//:TorchMLIRTMTensorDialect": [
```

GMNGeoffrey commented on third_party/torch-mlir-dialects/test/tmtensor/invalid.mlir:
```diff
@@ -0,0 +1,328 @@
+// RUN: iree-opt -split-input-file -verify-diagnostics %s
```
It seems like these have been modified? We don't really want to be modifying the code here. Amongst other things it makes the update process much more difficult. I don't think it's really necessary to have the tests here at all, actually. They aren't going to be included in our own build (we don't do this for any other third party dependency). If we're concerned about making sure we've got a good version of these files, we can run the tests in that repository before importing. If you just want a smoke test to make sure the dialect is registered and such, you can throw one in the main IREE repo (we can discuss best location for that). But also the tests you have for the pass should be able to test the thing we really care about, I think.

MaheshRavishankar replied:
I asked for these tests... (FTR, I think having this in `third_party/...` is not a great idea, but lets leave that aside). These are copy of the ops in IREE. Without having these tests they will silently fail and will only be known when Torch-MLIR gets bumped. I dont think that is a good setup but we have gone back and forth so long on this, I am not going to litigate this further. I'll just register my dissent on this and go away.

ScottTodd replied:
(I haven't looked at this specific example in detail but can take another more focused look after the conference this week)

Third party code must be included as pristine copies (or pristine copies + a clear list of patches) in third_party/.

We expect dependencies to be tested upstream. For the specific points where we use a dependency we test that behavior outside of third_party/. For example:
* https://github.com/google/iree/tree/main/iree/compiler/InputConversion/TOSA/test
* https://github.com/google/iree/tree/main/iree/test/e2e/tosa_ops

cathyzhyi replied:
Hey Scott, the tests I am trying to add are basic IR lit tests. Do you have any suggestions where to put them?

GMNGeoffrey reviewed (commented):
High-level comments:

It would be better to land this as multiple patches: one adding the third party dependency and a separate one adding the new code. We're going to need to do some work internally to get this working with the internal build and

cathyzhyi commented on iree/compiler/InputConversion/TMTensor/init_tm_tensor_dialect.h:
```diff
@@ -0,0 +1,27 @@
+// Copyright 2020 The IREE Authors
```
Done.

cathyzhyi commented on iree/compiler/InputConversion/TMTensor/init_tm_tensor_dialect.h:
```diff
@@ -0,0 +1,27 @@
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+// This files defines a helper to trigger the registration of dialects to
+// the system.
```
It seems all the other init_xx_dialects.h files contains this comment, like this one https://github.com/google/iree/blob/675e01231716801d8f5e3e91e89ae78857bfc655/iree/tools/init_iree_dialects.h#L7-L8.
Do we want to keep this for consistency?

ScottTodd replied:
Ah... keeping it consistent is fine, but I'll still grumble about it :P. We've had other boilerplate comments like "// Tests printing and parsing of ..." get spread throughout files, and eventually someone goes in and removes them, like with https://github.com/google/iree/pull/7968

cathyzhyi commented on iree/compiler/InputConversion/TMTensor/init_tm_tensor_dialect.h:
```diff
@@ -0,0 +1,27 @@
+
+// Add all the torch mlir dialects to the provided registry.
+inline void registerTMTensorDialect(DialectRegistry &registry) {
+  // clang-format off
+  registry.insert<mlir::torch::TMTensor::TMTensorDialect>();
+  // clang-format on
```
Removed the redundant `clang-format` hints as suggested. Thanks!

cathyzhyi commented on iree/compiler/InputConversion/init_tm_tensor_dialect.h:
```diff
@@ -0,0 +1,27 @@
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+// This files defines a helper to trigger the registration of dialects to
+// the system.
```
sry, this should have been removed.

cathyzhyi commented on iree/tools/init_dialects.h:
```diff
@@ -1,4 +1,4 @@
-// Copyright 2020 The IREE Authors
+// Copyright 2022 The IREE Authors
```
Reverted this change. Thanks!

cathyzhyi commented on iree/tools/init_dialects.h:
```diff
@@ -12,6 +12,7 @@
 #ifndef IREE_TOOLS_INIT_DIALECTS_H_
 #define IREE_TOOLS_INIT_DIALECTS_H_
 
+#include "iree/compiler/InputConversion/TMTensor/init_tm_tensor_dialect.h"
 #include "iree/tools/init_compiler_modules.h"
 #include "iree/tools/init_iree_dialects.h"
 #include "iree/tools/init_mlir_dialects.h"
```
I see. I will change it to follow the local style and change the file name to `init_torch_mlir_dialects.h` as well to be consistent.

stellaraccident commented:
Vacation snipped: these kind of IR "stubs" should just be copied into a project that needs them. Per policy, these need to be in third_party if originating "outside".
In hindsight, I may have put llvm-external-projects in third_party to better scalably comply with this bit of policy, but I don't think it is worth changing for this.
Please don't nit pick this and escalate to a very heavyweight (from an integration perspective) solution like submodules for such a thing. Policy compliance and minimizing llvm integrate pain should be the only thing in play here from a decision making standpoint.

ScottTodd commented:
> Vacation snipped: these kind of IR "stubs" should just be copied into a project that needs them. Per policy, these need to be in third_party if originating "outside". In hindsight, I may have put llvm-external-projects in third_party to better scalably comply with this bit of policy, but I don't think it is worth changing for this. Please don't nit pick this and escalate to a very heavyweight (from an integration perspective) solution like submodules for such a thing. Policy compliance and minimizing llvm integrate pain should be the only thing in play here from a decision making standpoint.

Fair points. I think most of my concerns go away by structuring this as one commit that adds the pristine import (with a readme explaining how to pull in updates), then any modifications to that code (including new cmake files) as separate git commits. The current state of the PR is pretty much there now, but going step by step from the start could have streamlined the review process.

stellaraccident commented:
> Fair points. I think most of my concerns go away by structuring this as one commit that adds the pristine import

Don't get me wrong: I am definitely in favor of getting the mechanics of complying with our policies right, and this is the first of this kind of dependency to be taken ("ir stubs") so is worth some discussion... It won't be the last and for there transient things, we need to be optimizing for developer throughput. The sooner in and working, the sooner the upstream shape of it takes, and the sooner we get it off our project specific books (which will play out over ~months of code movement). That's the real thing to optimize for here.

Ok, I'm out. Beach time :)

cathyzhyi commented on build_tools/third_party/torch-mlir-dialects/BUILD.overlay:
```diff
@@ -0,0 +1,151 @@
+load("@llvm-project//mlir:tblgen.bzl", "gentbl_cc_library", "td_library")
```
Added the license header. Thanks!

cathyzhyi commented on third_party/torch-mlir-dialects/README.md:
```diff
@@ -9,3 +9,12 @@ This project is intended to be used via LLVM's external projects setup:
 
 It depends on the `mlir` project.
+
+To pull the changes from https://github.com/llvm/torch-mlir:
```
Done.

ScottTodd commented:
> hmm, after rebasing to the head the `iree/test/e2e/matmul/e2e_matmul_direct_f32_large_split_k_dylib-llvm-aot_dylib` started to fail.

That's preexisting and should be fixed by https://github.com/google/iree/pull/8674.

GMNGeoffrey commented:
I can merge this past the failing test if you'd like (admin powers)

cathyzhyi commented:
> I can merge this past the failing test if you'd like (admin powers)

yes, please! Thanks a lot!

GMNGeoffrey commented:
Verified that it is only a single test failing in each of the workflows and it is a known failure. Merging past failing checks

GMNGeoffrey commented:
Also, merging with a merge commit, per my above comment that the commit history here is interesting and useful
